src/slurm_plugin/instance_manager.py

```diff
@@ -167,11 +167,20 @@ def _update_slurm_node_addrs(self, slurm_nodes, launched_instances):
                     print_with_count(zip(launched_nodes, launched_instances)),
                 )
             if fail_launch_nodes:
-                logger.warning(
-                    "Failed to launch instances due to limited EC2 capacity for following nodes: %s",
-                    print_with_count(fail_launch_nodes),
-                )
-                self._update_failed_nodes(set(fail_launch_nodes), "LimitedInstanceCapacity")
+                if launched_nodes:
+                    logger.warning(
+                        "Failed to launch instances due to limited EC2 capacity for following nodes: %s",
+                        print_with_count(fail_launch_nodes),
+                    )
+                    self._update_failed_nodes(set(fail_launch_nodes), "LimitedInstanceCapacity")
+                else:
+                    # EC2 Fleet doens't trigger any exception in case of ICEs and may return more than one error
+                    # for each request. So when no instances were launched we force the reason to ICE
+                    logger.error(
+                        "Failed to launch instances due to limited EC2 capacity for following nodes: %s",
```

Comment:
Just for my better understanding: in this way will we surface an ICE failure even if it may be something else?

Reply:
Yes.
CreateFleet has a very complex and rich behavior and in many cases where Run Instances will return an ICE, it just returns an error. Problem is that it may return a list of errors (triggered while attempting to provision different instances) but without information of how to match an error to an Instance.
As of now we do not have a mechanism to associate more than one error code to each instance so we agreed to consider an empty list of provisioned instances as a generic capacity issue (ICE) even if there may be different causes for different instances.
There is room for improvement but will require further investigation.